src/lib.rs

```diff
@@ -607,7 +614,17 @@ pub unsafe extern "C" fn fd_read(
         state.register_buffer(ptr, len);
 
         let read_len = u32::try_from(len).unwrap();
-        let file = state.get_file(fd)?;
+        let file = match state.get(fd)? {
+            Descriptor::File(f) => f,
+            Descriptor::Closed => return Err(ERRNO_BADF),
+            Descriptor::StdoutLog | Descriptor::StderrLog | Descriptor::Socket(_) => {
+                return Err(ERRNO_INVAL)
```

Comment:
This should fail with `ERRNO_ESPIPE`.

Reply:
This is an `fd_read` operation though which isn't performing a seek, so is `ESPIPE` the right error for this? (not that `INVAL` is any good, I'm not sure what posix does here)

Reply:
I'm sort of confused, though, this isn't a `pread` operation, just a `read`? Is `ESPIPE` used for `read`?

Reply:
Oh, oops. I somehow thought this was pread. It's BADF for reading from *Log, and socket can do whatever for now because we'll be implementing that soon.

Reply:
Updated!